### PR TITLE
[RFC] Adjust terminal text color if it is bold

### PR DESCRIPTION
**Pull request by DarkDefender**
It seems like neovim doesn't display the bold text with the same color as my other terminals do so I created this patch to fix this.

### CODE REVIEW

**Comment by DarkDefender** [comment]
Seems like I messed up the commit message. I guess I'll have to fix that before it is merged

**Comment by justinmk** [comment]
Would we want to do this even for terminals that support actual bold text? (I'm asking because I don't know. What is the common practice?)

**Comment by DarkDefender** [comment]
I'm not sure what you mean with "actual bold text". All the linux terminals that I've tried renders bold text that has one of the lower base colors (color 1-8) with the corresponding upper color (color 9-16). The text is still bold though so it will still render the text as bold.

For example my output of "ls" renders different filetypes with different colors. Some of the filetypes uses bold text.

I have some terminal applications that uses this "bold color shift method" also. I can provide screenshots if you want to.

I'm not sure what the comman practice is either. But this fix seems to make the neovim terminal match the text color rendering on my other terminals

**Comment by justinmk** [comment]
I had always thought that the color shift was an optional workaround for fonts (and terminals?) that don't support bold characters. I guess I was mistaken.

**Comment by cHoco** [comment]
@justinmk It is an option usually provided by the terminal emulator. This would effectively revert https://github.com/neovim/neovim/pull/3333

I personally think this change is wrong as per reason provided in the pull request linked above. We may need an option to configure the behaviour for GUI, but on terminal this shouldn't be a problem!

**Comment by DarkDefender** [comment]
@cHoco Thanks for pointing that out! I knew there must be some way to get it to use the default behaviour!

I guess we need to get to the bottom of why that option breaks you terminal, because without it, it breaks mine. See the attached screenshots:

Without reverting #3333:
without_revert
After reverting #3333
rever_3333

Notice that the bold colors doesn't render correctly in neovim without reverting #3333

**Comment by DarkDefender** [comment]
@cHoco Did you perhaps not define the nvim terminal colors with "g:terminal_color_[0-15]"?
That way you can just define the terminal colors 8-15 to be the same as 0-7.

Or neovim could switch off "bold_highbright" if the user only defined color 0-7. Or we could have it as an option.

As I said, this bold highbright thing seem to default to on on all the VTEs I tried on linux. Even xterm has this on per default.

**Comment by justinmk** [comment]
> Did you perhaps not define the nvim terminal colors with "g:terminal_color_[0-15]"?

Hmm, are you using `NVIM_ENABLE_TRUE_COLOR`? Those settings are only applicable for truecolor. 

Bold text and the shifted colors in nvim's `:terminal` _already_ work correctly (i.e., they respect the preferences I've set, in all forms) in both terminals I have: Terminal.app (OSX) and mintty (Windows).

**Comment by cHoco** [comment]
So did you define the `g:terminal_color_[0-15]` options? If you want the same colors inside and outside you shouldn't define them.
But this shows a clear need for an option deciding if the user wants to render bold text with an highlighted color, both when setting the terminal_color options manually or when using a GUI version of neovim the user can't decide this behaviour right now.

**Comment by DarkDefender** [comment]
@justinmk @cHoco ah, I see now! This is only a problem when running the neovim with NVIM_ENABLE_TRUE_COLOR.

With true color disabled it works as expected. However, because my terminal supports true color (and I want to use it) this is kind of a big problem for me.

If you don't want to add an additional option to toggle the bold_highbright, then perhaps just check if true color mode is on and switch it on then?

**Comment by justinmk** [comment]
`:term` just doesn't support truecolor yet. https://github.com/neovim/neovim/issues/2571 It is probably not hard to add support for it, depending on the capabilities of libvterm... someone just needs to do the work. Maybe you could look into it?

**Comment by DarkDefender** [comment]
@justinmk I did a quick hack to make the terminal inside neovim work with true color color codes.

However this doesn't really solve the need to have the bold_highbright option on...

**Comment by DarkDefender** [comment]
There, now I'm somewhat happy with the solution.

File: src/nvim/terminal.c [hunk 1]
```diff
@@ -217,10 +216,14 @@ Terminal *terminal_open(TerminalOptions opts)
   vterm_set_utf8(rv->vt, 1);
   // Setup state
   VTermState *state = vterm_obtain_state(rv->vt);
+  // Only relevant in terminal true color mode
+  if (true_color) {
```
Comment (cHoco) [inline src/nvim/terminal.c]: I know we don't like adding new configuration vars, but instead of enabling by default if true color is enabled, personally, I'd prefer a new option like 'g:terminal_color_[0-15]' to control this behaviour.

Comment (justinmk) [inline src/nvim/terminal.c]: Yeah, I'm just trying to think of where to put these options, since they are starting to take root.

**Comment by DarkDefender** [comment]
I guess we can keep the bold_highbright on by default in true color mode till you guys decide if you want to add a new option or not.

However, it seem like I have one test failing now. I don't know if the test needs to be updated or if there really is a problem with the new code. All my quick tests in my terminals doesn't seem to break the colors in neovim term (with and without true color)

**Comment by justinmk** [comment]
> I guess we can keep the bold_highbright on by default in true color mode till you guys decide if you want 

That's fine with me. Please squash and fix the commit message. Though I am not sure about the change you made to the test.

**Comment by DarkDefender** [comment]
@justinmk The test failed previously because it tried to set the terminal "g:terminal_color" and check the result when it wasn't in true color mode. So I made sure that it would be in true color mode when it checks if the correct rgb color is printed.

I guess that we should perhaps reset it to non true color mode after the test is done?

**Comment by DarkDefender** [comment]
should I mark this as RDY or is there anything else I need to fix?

File: test/functional/terminal/highlight_spec.lua [hunk 2]
```diff
@@ -134,7 +134,7 @@ describe('terminal window highlighting with custom palette', function()
     })
     screen:attach(true)
     nvim('set_var', 'terminal_color_3', '#123456')
-    execute('enew | call termopen(["'..nvim_dir..'/tty-test"]) | startinsert')
+    execute('enew | let $NVIM_TUI_ENABLE_TRUE_COLOR=1 | call termopen(["'..nvim_dir..'/tty-test"]) | startinsert')
```
Comment (justinmk) [inline test/functional/terminal/highlight_spec.lua]: This should not be necessary; in the `screen:attach(true)` call above, `true` specifies that the attached UI is an RGB ("true color") UI. That's what you should check in `terminal_open()`, not `os_getenv("NVIM_TUI_ENABLE_TRUE_COLOR") != NULL`

File: src/nvim/terminal.c [hunk 3]
```diff
@@ -250,12 +253,16 @@ Terminal *terminal_open(TerminalOptions opts)
...
 
+  if (!true_color) {
+      return rv;
+  }
+
   // Configure the color palette. Try to get the color from:
   //
   // - b:terminal_color_{NUM}
   // - g:terminal_color_{NUM}
   // - the VTerm instance
```
Comment (justinmk) [inline src/nvim/terminal.c]: the lines removed below mean that "the VTerm instance" is no longer queried for default colors. Why was that removed? And if we go with that, then this comment needs to be updated to reflect that.

**Comment by justinmk** [comment]
@DarkDefender Looks like you simplified some things but there is no explanation. Can you help me understand the changes?

File: src/nvim/terminal.c [hunk 4]
```diff
@@ -205,6 +202,8 @@ void terminal_teardown(void)
 
 Terminal *terminal_open(TerminalOptions opts)
 {
+  // Will this be a true color terminal?
+  bool true_color = os_getenv("NVIM_TUI_ENABLE_TRUE_COLOR") != NULL;
```
Comment (justinmk) [inline src/nvim/terminal.c]: use `ui_rgb_attached()` (from ui.h) instead.

    bool true_color = ui_rgb_attached();

File: src/nvim/terminal.c [hunk 5]
```diff
@@ -250,12 +253,16 @@ Terminal *terminal_open(TerminalOptions opts)
   rv->sb_size = MIN(rv->sb_size, 100000);
   rv->sb_buffer = xmalloc(sizeof(ScrollbackLine *) * rv->sb_size);
 
+  if (!true_color) {
+      return rv;
```
Comment (justinmk) [inline src/nvim/terminal.c]: 2-space indent

File: test/functional/terminal/highlight_spec.lua (hunk 2 above)
Comment (DarkDefender) [inline test/functional/terminal/highlight_spec.lua]: Ah, ok. I guess this will be solved if I use ui_rgb_attached() as you suggested below.

File: src/nvim/terminal.c (hunk 3 above)
Comment (DarkDefender) [inline src/nvim/terminal.c]: I did this because it you are not in true color mode you already know which color id to output (they are stored and inited into color_indexes).

And because the g:terminal_color colors wouldn't be used if we are not in true color mode, I simply skipped over them.

I also now use the native libvterm way to set the default colors. This is a lot better and fixes the problem where a true color code would be incorrectly converted to an indexed color. 

However because I use the native way to set colors, it will mess up the color_indexes map. This will only become a problem in none true color mode though, so that is why I added the return.

I'm not sure how I would write all that into a comment though... Is "Only set custom terminal colors if in true color mode" enough?

File: src/nvim/terminal.c (hunk 4 above)
Comment (DarkDefender) [inline src/nvim/terminal.c]: Will that tell me if NVIM_TUI_ENABLE_TRUE_COLOR is set or not? I mean if it only tells me if the terminal is capable of true color, then it is not enough as it will break index colors if the user sets g:terminal_color colors and has not enabled NVIM_TUI_ENABLE_TRUE_COLOR.

Comment (justinmk) [inline src/nvim/terminal.c]: NVIM_TUI_ENABLE_TRUE_COLOR is a hack that tells the TUI to attach as a "rgb enabled UI". NVIM_TUI_ENABLE_TRUE_COLOR is really not relevant here. The authority is the UI.

File: src/nvim/terminal.c (hunk 3 above)
Comment (justinmk) [inline src/nvim/terminal.c]: > Is "Only set custom terminal colors if in true color mode" enough?

Don't bother; the if-statement communicates that clearly enough.

I'm not sure about color_indexes, I'll take your word for it. Does this have any bearing on the following:

https://github.com/neovim/neovim/issues/3459

https://github.com/neovim/neovim/issues/2323

Comment (DarkDefender) [inline src/nvim/terminal.c]: Those two problems won't be solved by this patch if it's what you are asking.

However, it seems like the color_index map might be the cause for those problems. As noted in #2323 there will be some color collisions in the map and the color with the highest index will be returned.

Because it would be best to just return the true index and not just an index that had the same color, perhaps we should remove the color_index map completely.

Instead we will (if the ui is not true color) init all the 256 default colors to (0,0,[0-255]). This should produce an int that is [0-255] so we can the just pass the value to the index value with no lookup needed.

File: src/nvim/terminal.c [hunk 6]
```diff
@@ -205,6 +201,8 @@ void terminal_teardown(void)
 
 Terminal *terminal_open(TerminalOptions opts)
 {
+  // Will this be a true color terminal?
```
Comment (justinmk) [inline src/nvim/terminal.c]: Comment is redundant, consider removing it

**Comment by justinmk** [comment]
@tarruda can you sanity check this?

**Comment by DarkDefender** [comment]
@justinmk I have implemented the solution to #2323 that I suggested.
Do you want me to open a new PR for that one or should I add it to this one?

**Comment by justinmk** [comment]
@DarkDefender Nice! Let's say separate PR so we can move forward with this one.

File: src/nvim/terminal.c [hunk 7]
```diff
@@ -217,10 +214,14 @@ Terminal *terminal_open(TerminalOptions opts)
   vterm_set_utf8(rv->vt, 1);
   // Setup state
   VTermState *state = vterm_obtain_state(rv->vt);
+  // Only relevant in terminal true color mode
+  if (true_color) {
+    vterm_state_set_bold_highbright(state, true);
```
Comment (justinmk) [inline src/nvim/terminal.c]: @DarkDefender Is there a reason this needs to be done here, instead of after the `if (!true_color)` check below?

**Comment by DarkDefender** [comment]
@justinmk It depends on the changes made in this PR though... So I guess I have to wait till this one is merged then?

**Comment by justinmk** [comment]
@DarkDefender Oh, make it a separate commit on this one then.

File: src/nvim/terminal.c (hunk 7 above)
Comment (DarkDefender) [inline src/nvim/terminal.c]: I don't think it has to be there. It is just there because #3333 had it there.

I didn't move it down because I didn't know if you wanted to change the check to be something else later.

I can move it down now if you want to.

Comment (justinmk) [inline src/nvim/terminal.c]: Oh, right. In order to not regress #3333, let's add a `g:terminal_bold_highbright` setting. 

Though your suggestion at https://github.com/neovim/neovim/pull/4274#issuecomment-185291835 avoids a new setting, it will require more code and more documentation to explain it. 

Avoiding another `g:terminal_` option right now doesn't buy us much until all of the `g:terminal` options are replaced with some other scheme, so, adding a new one makes no difference.

Comment (DarkDefender) [inline src/nvim/terminal.c]: but wouldn't `g:terminal_bold_highbright` only be used in true color mode anyways?

I mean, in non true color mode there is no reason to have highbright on because the user terminal does the color shifting anyways.

But perhaps you want me to move it down to the below the `(!true_color)` check and have an additional check for `g:terminal_bold_highbright`?

Comment (justinmk) [inline src/nvim/terminal.c]: As long as we aren't regressing #3333 it does not matter to me. If we can avoid the option, great. @cHoco does this work for you?

Comment (cHoco) [inline src/nvim/terminal.c]: From reading the code I didn't get the idea that the `g:terminal` options were used only in true color mode (they supported 256 colors). I personally wasn't using but it's something to take into account. 
An option `g:terminal_bold_highbright` anyway is needed, and I wouldn't put it behind a true color mode check since it would be relevant in both cases.

Comment (DarkDefender) [inline src/nvim/terminal.c]: No, if you are talking about the `g:terminal_color` options then those were only used in true color mode. They were never used to change the color index. They only changed the rgb (true color value) of the cell.

Anyways onto the highbright stuff. The reason I have it as it is is because:

1. In non true color more you already have it off because you want to respect the user terminal setting regarding highbright. I can see the having an option for it only if you have highbright off in your normal terminal and want to have it on in neovim terminal.

However, this setting will only work one way. That is, if the user terminal has highbright on, you can't turn it off in neovim. So this option could be kinda confusing in non true color mode.

2. In true color mode this should probably be on by default. As I stated previously, it seems like most if not all linux terminals defaults to highbright on. So I would say that the expected behaviour of the neovim terminal would be to have it on.

This will clash with the default to off in non true color mode.

So I think that the only sane way to do this is to have the option only when in true color mode. It's also a bit more logical because then all the `g:terminal_` options are only used in true color mode.

Comment (justinmk) [inline src/nvim/terminal.c]: @DarkDefender Thanks for covering all of those scenarios. It is clear now we don't need the extra setting.

Comment (DarkDefender) [inline src/nvim/terminal.c]: Moved the highbright option to below the `(!true_color)` as you suggested

**Comment by DarkDefender** [comment]
I did take a look at the libvterm code base and i noticed that you can not change any palette colors besides the first 16. So I changed the for loop to only init the first 16 colors.

File: src/nvim/terminal.c (hunk 7 above)
Comment (cHoco) [inline src/nvim/terminal.c]: @DarkDefender That makes sense, putting the option `g:terminal_bold_highbright` after the true color check seems reasonable. I wouldn't be able to fake highlight off because terminal application themes I use use all the 16 base colours.

Comment (justinmk) [inline src/nvim/terminal.c]: @cHoco Are you using truecolor?

Comment (cHoco) [inline src/nvim/terminal.c]: @justinmk Not in terminal, using the the rgb interface when I occasionally use a neovim gui (nyaovim)
